core/domain/exp_jobs_one_off.py

```diff
@@ -597,6 +601,62 @@ def reduce(status, values):
             yield (status, values)
 
 
+class ImagesAuditJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for checking whether all images in {{exp_id}}/assets/image/
+    are also present in exploration/{{exp_id}}/assets/image/.
+    """
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(exp_model):
+        if exp_model.deleted:
+            return
+
+        if not constants.DEV_MODE:
+            exp_id = exp_model.id
+            fs_internal = fs_domain.AbstractFileSystem(
+                fs_domain.GcsFileSystem(exp_id))
+            fs_external = fs_domain.AbstractFileSystem(
+                fs_domain.GcsFileSystem('exploration/' + exp_id))
+            # We have to make sure we pass the dir name without starting or
+            # ending with '/'.
+            image_urls_internal = fs_internal.listdir('image')
+            image_urls_external = fs_external.listdir('image')
+
+            image_filenames_internal = set([
+                GCS_IMAGE_ID_REGEX.match(url).group(3)
+                for url in image_urls_internal])
+
+            image_filenames_external = set([
+                GCS_EXTERNAL_IMAGE_ID_REGEX.match(url).group(3)
+                for url in image_urls_external])
+
+            # Currently in editor.py and exp_services.py, all images are stored
+            # in exploration/{{exp_id}}/ parent folder. So, we only need to
+            # check if all images in the internal folder are there in the
+            # external folder.
+            non_existent_images = image_filenames_internal.difference(
+                image_filenames_external)
+
+            if len(non_existent_images) > 0:
+                yield (
+                    exp_id,
+                    'Missing Images: %s' % list(non_existent_images))
+            else:
+                yield (
+                    ALL_IMAGES_VERIFIED,
+                    'All images verified for exploration: %s' % (exp_id))
```

Comment:
I think, here, let's just add a count of the total number of image filenames verified. Then, reduce() can sum those counts.

This is to catch cases where for some reason we are checking the wrong thing and the count is 0. 

Note that we don't need to store the exp IDs at all (for the success case).

Reply:
Done